From 3ee26faa201cc0f6dd4ef443b11bfdb39991ab9c Mon Sep 17 00:00:00 2001
From: Xianzhong <xianzhong.li@nxp.com>
Date: Mon, 4 Jun 2018 23:51:23 +0800
Subject: [PATCH 3941/5242] MGS-3940-2 gpu: imx: dpu-blit: fix event trigger

commit  97c8ae0570031d098a53d72a8e6c316ece929d32 from
https://source.codeaurora.org/external/imx/linux-imx.git

the previous programming does not flush command sequence,
that cause the obvious flicker when run glmark2 in full-screen,
also the wrong logic cause wayland hang in strict conditions.

need program event trigger to sync command sequencer properly.
also revert workaround: 4767f9aaad6d8167c21b329ace00afd3a01cd336
MLK-18283-3 gpu: imx: dpu-blit: sync for cmd sequence finished

Signed-off-by: Xianzhong <xianzhong.li@nxp.com>
Signed-off-by: Meng Li <Meng.Li@windriver.com>
---
 drivers/gpu/imx/dpu-blit/dpu-blit.c |    8 +++++---
 1 file changed, 5 insertions(+), 3 deletions(-)

diff --git a/drivers/gpu/imx/dpu-blit/dpu-blit.c b/drivers/gpu/imx/dpu-blit/dpu-blit.c
index 8c2c5f8..242b3a8 100644
--- a/drivers/gpu/imx/dpu-blit/dpu-blit.c
+++ b/drivers/gpu/imx/dpu-blit/dpu-blit.c
@@ -121,8 +121,6 @@ void dpu_be_configure_prefetch(struct dpu_bliteng *dpu_be,
 	 */
 	if (tiled_work_unfinished || baddr) {
 		dpu_be_wait(dpu_be);
-		dpu_cs_wait_idle(dpu_be);
-		udelay(10);
 		tiled_work_unfinished = false;
 	}
 
@@ -249,7 +247,11 @@ int dpu_be_blit(struct dpu_bliteng *dpu_be,
 #define STORE9_SEQCOMPLETE_IRQ_MASK	(1U<<STORE9_SEQCOMPLETE_IRQ)
 void dpu_be_wait(struct dpu_bliteng *dpu_be)
 {
-	dpu_be_write(dpu_be, 0x10, PIXENGCFG_STORE9_TRIGGER);
+	dpu_cs_wait_fifo_space(dpu_be);
+
+	dpu_be_write(dpu_be, 0x14000001, CMDSEQ_HIF);
+	dpu_be_write(dpu_be, PIXENGCFG_STORE9_TRIGGER, CMDSEQ_HIF);
+	dpu_be_write(dpu_be, 0x10, CMDSEQ_HIF);
 
 	while ((dpu_be_read(dpu_be, COMCTRL_INTERRUPTSTATUS0) &
 		STORE9_SEQCOMPLETE_IRQ_MASK) == 0)
-- 
1.7.9.5

